packages/components/src/utils/hooks/use-controlled-value.ts

```diff
@@ -27,8 +27,18 @@ export function useControlledValue< T >( {
 	const initialValue = hasValue ? valueProp : defaultValue;
 	const [ state, setState ] = useState( initialValue );
 	const value = hasValue ? valueProp : state;
-	const setValue =
-		hasValue && typeof onChange === 'function' ? onChange : setState;
+
+	let setValue: ( nextValue: T ) => void;
+	if ( hasValue && typeof onChange === 'function' ) {
+		setValue = onChange;
+	} else if ( ! hasValue && typeof onChange === 'function' ) {
+		setValue = ( nextValue ) => {
+			onChange( nextValue );
+			setState( nextValue );
+		};
+	} else {
+		setValue = setState;
+	}
```

Comment:
Does this "pattern" have a standard name (semi-controlled components)? It'd be nice to include it in future documentation with a brief explanation of its working principles (cc @diegohaz )

Reply:
It's not controlled unless you're passing a `value` prop, so semi-controlled seems misleading.

This just mirrors more closely how the intrinsic `input` controls work. I think it still falls under the controlled/uncontrolled dichotomy?

Reply:
Gotcha. I think I'm not going to mention it necessarily in the guidelines, for now